src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingListObjectsOperation.cs

```diff
@@ -0,0 +1,219 @@
+using Microsoft.SqlTools.Utility;
+using Microsoft.SqlTools.Hosting.Protocol.Contracts;
+
+namespace Microsoft.SqlTools.ServiceLayer.Scripting
+{
+    /// <summary>
+    /// Class to represent an in-progress list objects operation.
+    /// </summary>
+    public sealed class ScriptingListObjectsOperation : ScriptingOperation
+    {
+        private CancellationTokenSource cancellation = new CancellationTokenSource();
+
+        private bool disposed = false;
+
+        private int scriptedObjectCount = 0;
+
+        private int totalScriptedObjectCount = 0;
+
+        private ScriptingListObjectsParams Parameters { get; set; }
+
+        private RequestContext<ScriptingListObjectsResult> RequestContext { get; set; }
+
+        public ScriptingListObjectsOperation(ScriptingListObjectsParams parameters, RequestContext<ScriptingListObjectsResult> requestContext)
+        {
+            if (parameters == null)
+            {
+                throw new ArgumentNullException("parameters");
+            }
+
+            if (requestContext == null)
+            {
+                throw new ArgumentNullException("requestContext");
+            }
+
+            this.OperationId = Guid.NewGuid().ToString();
+            this.Parameters = parameters;
+            this.RequestContext = requestContext;
+        }
+
+        public Task ActiveTask { get; private set; }
+
+        public override Task Execute()
+        {
+            string operationId = Guid.NewGuid().ToString();
+            this.ActiveTask = Task.Run(() => this.InternalExecute(), this.cancellation.Token);
+            return this.ActiveTask;
+        }
+
+        private void InternalExecute()
+        {
+            if (this.cancellation.IsCancellationRequested)
+            {
+                throw new OperationCanceledException(this.cancellation.Token);
+            }
+
+            SqlScriptPublishModel publishModel = null;
+
+            try
+            {
+                this.ValidateScriptDatabaseParams();
+
+                publishModel = new SqlScriptPublishModel(this.Parameters.ConnectionString);
+                publishModel.ScriptItemsCollected += this.OnPublishModelScriptItemsCollected;
+                publishModel.ScriptProgress += this.OnPublishModelScriptProgress;
+                publishModel.ScriptError += this.OnPublishModelScriptError;
+
+                List<ScriptingObject> databaseObjects = publishModel.GetDatabaseObjects();
+
+                ScriptingListObjectsCompleteParameters eventParameters = new ScriptingListObjectsCompleteParameters
+                {
+                    OperationId = this.OperationId,
+                    DatabaseObjects = databaseObjects,
+                    Count = databaseObjects.Count,
+                };
+
+                this.SendJsonRpcEventAsync(ScriptingListObjectsCompleteEvent.Type, eventParameters);
+            }
+            catch (Exception e)
+            {
+                if (e.IsOperationCanceledException())
```

Comment:
This extension method checks inner exceptions - SMO wraps all exceptions in it's own exception class.

---
In reply to: [111653257](https://github.com/Microsoft/sqltoolsservice/pull/316#discussion_r111653257) [](ancestors = 111653257)